Add Host key verification failed page (#25401)
github/docs: content/authentication/troubleshooting-ssh/host-key-verification-failed.md

---
title: Host key verification failed
intro: 'As a security precaution, ssh keeps track of which hosts it has previously seen.'
versions:
  fpt: '*'
  ghec: '*'
topics:
  - SSH
shortTitle: Host key verification failed
---

> This means that the key that the server to which your client is connecting presented a key that doesn't match the keys that your client has seen in the past.

Reviewer: There's some duplication here

Is there a way we can use language that is slightly more beginner-friendly than 'client' here? For example, could we just say "the server to which you're connecting"? Alternatively, can we be slightly more explicit as to what we mean (e.g. "your SSH client")?

```suggestion
This error means that the server to which your client is connecting presented a key that doesn't match the keys that your client has seen in the past.
```

Author: So, `client` here on average, more or less, means `git`. In general that's effectively `git.exe` or `/usr/bin/git`, although if someone is using [`jgit`](https://git-scm.com/book/uz/v2/Appendix-B%3A-Embedding-Git-in-your-Applications-JGit), or [go-git](https://github.com/go-git/go-git) or [dulwich](https://github.com/jelmer/dulwich) (which is sometimes used, e.g. by Mercurial to talk to git via [HgGit](https://wiki.mercurial-scm.org/HgGit)), [libgit2](https://github.com/libgit2/libgit2), or [js-git](https://github.com/creationix/js-git), it could mean one of those.

But, it could mean any [ssh implementation](https://ssh-comparison.quendi.de/) in case someone is actually just connecting directly via `ssh`.

Author: Basically, I'm using `client` to cover for the fact that I don't know if they're using `git` (or an implementation) or `ssh` (or an implementation).

Author: And, if they're using `git` (or an implementation) or worse a wrapper for git, like an IDE -- whether that's IntelliJ, VSCode, Eclipse, or something else) -- I can't know that they'll really know what `ssh` is or that they're using it.

Reviewer: I think "the server to which you're connecting" might be a useful abstraction here, but again go with your preference

Author: I'll take that, but it still leaves a `client` at the end of the sentence.

One might consider:
```diff
-doesn't match the keys that your client has seen in the past.
+doesn't match the keys that it has used in the past.
```

But this could easily be a globally incorrect statement while technically accurate from the perspective of the local client. And I'm not terribly comfortable w/ that tradeoff. (Can you tell I like my precision?) 

But, let's try a variant of that and see what you think.

> It could be because the server changed its keys unexpectedly -- this has happened -- in which case there should be a news reports from trustworthy sources (including the server's official blog) announcing the change.

Reviewer: I think it would be useful to keep this article at least partly focused on GitHub, and explain how _GitHub_ handles changing its keys.

```suggestion
You may see this error if the server has changed its keys unexpectedly, in which case you should be able to find an official report from a trustworthy source announcing the change. If {% data variables.product.company_short %} changes its SSH host key, this will be announced on {% data variables.product.prodname_blog %} at [github.blog](https://github.blog/).
```

> 

Reviewer: I think it's worth mentioning the case where the server hasn't actually recently changed, but the key isn't in your known hosts file for whatever reason. Again we can keep this GitHub focused.

```suggestion

You can find an up-to-date list of {% data variables.product.company_short %}'s public SSH keys on {% data variables.product.prodname_docs %}. You may need to add these keys to your `~/.ssh/known_hosts` file. For more information, see "[AUTOTITLE](/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints)."

```

Author: I should note that `~/.ssh/known_hosts` is _mostly_ a Unix-ism. But I don't have the energy to think about an alternative suggestion (this is mostly a problem for Windows where `~` is mostly not a thing).

I'm going to take this suggestion, and basically each suggestion where I added 👍 and 👀 -- as a pair the 👀 are a reminder to me to think about and the 👍 is a thing I like -- once I've incorporated a thing, I'll remove the 👀.

Reviewer: That's a good point, feel free to simplify to "`known_hosts` file"! Id we wanted to make that change then I think the article we're linking to should do the heavy lifting of explaining the file's location across different OSs

> It could also be that for one reason or another your computer is connecting to a server that is **not** the the right server. In such cases, you probably do not want to connect. You'll probably want to reach out to your support, and then the server's support for help. If the server is being impersonated, the owner of the server will probably appreciate being informed.

Reviewer: I think we can use more direct language here. "In such cases, you probably do not want to connect" is not necessarily as useful as being able to identify what these cases are.

We use "computer" here but "client" above. Is there any reason for this difference? Again, can we just simplify to "you are connecting"?

```suggestion
If you are encountering the error but can't find an official source for the server's keys, it is safest not to connect, because you may be trying to connect to a server that is not the one you intend. You may want to contact your IT department or the server's support team for help. If the server is being impersonated, the owner of the server will appreciate being informed.
```

Author: I think (it was a while ago) I used `computer` because while the client initiated the connection, there are lots of things outside the client that will impact the connection path, and it ends up being the computer that makes the connection to the world. DNS for instance is generally controlled at the computer level, and where the computer physically is does too. 

If that distinction isn't important enough, I'm not terribly attached to it.

Reviewer: I'd just go with something like "...you are connecting..." personally. I think it's best to think in terms of what users are doing/trying to do. Talking about the computer connecting to an unintended server sounds quite passive and might even instil panic (like there's something wrong with the computer itself 😁)

Reviewer: See what you think and give me a ping when you're ready for me to take another look 🙂

Author: Well, there *could* be something wrong with the computer itself! 😄

But, yeah, lemme think about it for a bit. In general, I'm not terribly attached to text -- I prefer incremental improvements even if they aren't my initial wording.

Reviewer: I think `being informed` works there! 😁 Because the reader is still the implied agent. I'm not against the passive voice in general, just think it's worth keeping focused on what the reader can/should do. Anyway we might be getting too in the weeds here, so go with whatever you're happy with

> For more information, see "[GitHub's SSH Key Fingerprints](/authentication/keeping-your-account-and-data-secure/githubs-ssh-key-fingerprints)."

Reviewer: I'm suggesting we reference this article above instead.

```suggestion
```
